lib/packwerk/application_load_paths.rb

```diff
@@ -18,30 +18,28 @@ def extract_relevant_paths(root, environment)
         relative_path_strings(relevant_paths)
       end
 
-      sig { returns(T::Hash[String, Module]) }
+      sig { returns(T::Array[String]) }
       def extract_application_autoload_paths
-        Rails.autoloaders.inject({}) do |h, loader|
-          h.merge(loader.root_dirs)
-        end
+        Zeitwerk::Loader.all_dirs
```

Comment:
I'm willing to help in new use cases arise. For example, if a read-only collection of root directories with their associated namespaces would be a good solution for what this is trying to solve, I'd be open to implement it.

However, I don't know if this is a good solution to the root problem. If a gem is not an engine, and does not use Zeitwerk either, what?

Take also into account that Zeitwerk is used by gems. How are you going to distinguish the loaders that matter among them?

Reply:
@mclark it feels like we should have used packwerk to define which parts of packwerk are public API and which are not ;) jkjk

Do you mind elaborating as to what you do with the default namespace?

Thanks for offering to support this, Xavier! To answer your questions:

> If a gem is not an engine, and does not use Zeitwerk either, what?

Such gems would currently not be included. And I think that is fine because...

> Take also into account that Zeitwerk is used by gems. How are you going to distinguish the loaders that matter among them?

... currently packwerk only operates on constants defined with the Rails application. I.e., packwerk is focussed on helping engineers understand and clean up the dependency structures within their own applications, so there is already a boundary built in. https://github.com/Shopify/packwerk/pull/216 is looking to change this and add more options, but the fundamental fact that packwerk has controls for this won't change.

How is any of this useful? One example of an app that uses gems and engines as part of the codebase is https://github.com/instructure/canvas-lms/tree/master/gems, which defines a bunch of gems inline and uses the directly from the Gemfile (https://github.com/instructure/canvas-lms/blob/master/Gemfile.d/app.rb#L160)

Reply:
@shageman sorry I missed your comment somehow!

> Do you mind elaborating as to what you do with the default namespace?

Some legacy directories in GitHub do not fully conform to the patterns established by Zeitwerk and have an implicit "root" namespace. When we migrated to Zeitwerk we leveraged the ability to set a default namespace for each root directory in order to override the default `Object` root in these cases.

For Packwerk to resolve constants properly, it too needs to know what the default root namespaces are otherwise the expected and actual namespaces for each file will not match.

This isn't an ideal situation, but for the foreseeable future we will be stuck with this pattern unfortunately. Thanks for taking the time to understand the problem.

Reply:
Let me say that I believe root namespaces are as good practice as `Object`. If I find an API that I like and is backwards compatible, I'd like to provide configuration for them in Rails somehow. Nowadays, you can do this reaching for the autoloaders by hand, but I believe this should be better supported.

I have been consulting for many years, and it is rare the project that does not have `app` in the autoload paths. That is a weird configuration, and one that is forced because teams just one `app/api` to be `API`, for example. The fact that `app/models` is not a namespace does not mean people do not want `app/components` to be one, because we'll all agree `app/components/components` is screaming the current settings are not sufficient.